lib/clone.js

```diff
@@ -23,125 +25,157 @@ function clone(pkg, src, dst, credentials, xform, ready) {
 
   if(arguments.length === 5) {
     ready = xform
-    xform = null
+    xform = through()
   } else if(arguments.length === 4) {
     ready = credentials
-    xform = credentials = null
+    credentials = null
+    xform = through()
   }
 
-  var metadataUrl = url.resolve(
-      sourceRegistry
-    , pkg.name + '/' + (pkg.version || '*')
-  )
-
   var shasum = crypto.createHash('sha1')
-    , parse = tar.Parse()
+    , transformedTarball = tar.pack()
+    , readmeFilename
+    , packageHeader
+    , newVersion
+    , readmeData
     , metadata
-    , tarball
     , publish
-    , parsed
-    , setup
-    , dist
-    , sha
-    , req
 
-  req = request(metadataUrl)
+  var originalTarball = fetch(pkg.name, pkg.version)
+    .on('error', onerror)
 
-  req
-      .on('error', onerror)
-    .pipe(xform || through())
+  transformedTarball.pipe(concat(gzip))
```

Comment:
so I was doing something dumb, not sure what I was doing exactly, but I rewrote this to be a little cleaner and not use `.on('data')` anywhere and seems like the race condition I was seeing disappeared. Anyway, no more concat-streaming the whole thing before gziping